### ISSUE TITLE
"Match_: Do not compare objects directly" error incorrectly reported for `$this` usage in `match` when `$this` is an enum

### ISSUE THREAD

**Issue by ZebulanStanphill** [voku/phpstan-rules#42]
#### What is this feature about (expected vs actual behaviour)?

A "Match_: Do not compare objects directly" error is incorrectly reported when using a `match` statement in an enum method like so...

(I think it has something to do with the usage of `$this`, which happens to refer to an enum in this particular use-case.

#### How can I reproduce it?

```php
enum MembershipLevel {
	case Free;
	case Standard;
	case Premium;
	case PremiumPlus;

	public function isPaid(): bool {
		return match ($this) {
			// This makes voku/phpstan-rules unhappy.
			self::Free => false,
			self::Standard, self::Premium, self::PremiumPlus => true,
		};
	}
}
```

#### Does it take minutes, hours or days to fix?

Hours, I would assume, but I'm not familiar with your codebase.

**Comment by spenserhale** [comment]
+1

Also experiencing this false positive.
